src/cosim/utility/thread_pool.hpp

```diff
@@ -0,0 +1,106 @@
+/**
+*  \file
+*  Slave interface.
+*
+*  \copyright
+*      This Source Code Form is subject to the terms of the Mozilla Public
+*      License, v. 2.0. If a copy of the MPL was not distributed with this
+*      file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+#ifndef COSIM_UTILITY_THREAD_POOL_HPP
+#define COSIM_UTILITY_THREAD_POOL_HPP
+
+#include <atomic>
+#include <condition_variable>
+#include <functional>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+namespace cosim
+{
+namespace utility
+{
+
+class thread_pool
+{
+private:
+    std::atomic_bool done_;
+    std::queue<std::function<void()>> work_queue_;
+    std::vector<std::thread> threads_;
+    std::mutex m_;
+    std::condition_variable cv_;
+
+    void worker_thread()
+    {
+        while (!done_) {
+
+            std::unique_lock<std::mutex> lck(m_);
+            if (!work_queue_.empty()) {
+
+                auto task = work_queue_.front();
+                work_queue_.pop();
+                task();
```

Comment:
This might causing the simulation (fmu steps) not parallelized since other worker threads are blocked until the mutex gets released.

Reply:
I did not create a separate PR but I see the simulation runs faster with the following modification:
```patch
Index: src/cosim/utility/thread_pool.hpp
<+>UTF-8
===================================================================
diff --git a/src/cosim/utility/thread_pool.hpp b/src/cosim/utility/thread_pool.hpp
--- a/src/cosim/utility/thread_pool.hpp	(revision HEAD)
+++ b/src/cosim/utility/thread_pool.hpp	(revision Staged)
@@ -30,23 +30,27 @@
     std::vector<std::thread> threads_;
     std::mutex m_;
     std::condition_variable cv_;
+    std::condition_variable wk_cv_;
+    uint32_t num_tasks;
 
     void worker_thread()
     {
+        std::chrono::seconds wait_duration(1);
         while (!done_) {
-
             std::unique_lock<std::mutex> lck(m_);
+
             if (!work_queue_.empty()) {
-
                 auto task = work_queue_.front();
                 work_queue_.pop();
-                task();
-
                 lck.unlock();
+                task();
+
+                lck.lock();
+                num_tasks--; // This needs to be decremented after the task is actually completed
                 cv_.notify_one();
-
             } else {
                 std::this_thread::yield();
+                wk_cv_.wait_for(lck, wait_duration);
             }
         }
     }
@@ -54,6 +58,7 @@
 public:
     explicit thread_pool(unsigned int thread_count = std::thread::hardware_concurrency())
         : done_(false)
+        , num_tasks(0)
     {
         try {
             for (unsigned i = 0; i < thread_count; ++i) {
@@ -73,7 +78,9 @@
     {
         if (!threads_.empty()) {
             std::unique_lock<std::mutex> lck(m_);
-            while (!work_queue_.empty()) cv_.wait(lck);
+            while (num_tasks > 0) {
+                cv_.wait(lck);
+            }
         }
     }
 
@@ -84,14 +91,16 @@
             f();
         } else {
             std::unique_lock<std::mutex> lck(m_);
+            num_tasks++;
             work_queue_.push(std::function<void()>(f));
+            wk_cv_.notify_one();
         }
     }
 
     ~thread_pool()
     {
         done_ = true;
-
+        wk_cv_.notify_all();
         for (auto& thread : threads_) {
             if (thread.joinable()) {
                 thread.join();

```

Reply:
Adding my suggestions in code as well (may include typos since I just wrote this into the text editor);

Changes to `worker_thread()`:
```
void worker_thread()
{
    while (true)
    {
        {
            std::unique_lock<std::mutex> lck(m_);
            cv_.wait(lck, [=]{return done || !work_queue_.empty()});
    
            if (done && work_queue_.empty()) break; 

            auto task = std::move(work_queue_.front());
            work_queue_.pop();
        }
        task();
    }
}
```

Changes to `submit()`:

```
template<typename FunctionType>
void submit(FunctionType f)
{
    if (threads_.empty()) {
        f();
    } else {
       {
           std::unique_lock<std::mutex> lck(m_);
           work_queue_.emplace(std::move(std::function<void()>(f)));
       }
      cv_.notify_one();
   }
}
```